[Themes] Add new semantic color scheme

romoore opened the pull request:
The new semantic color scheme includes a number of new properties focused on their contextual use within components. This should make it easier to provide consistent, holistic theming for an entire app using a set of baseline colors and mappings.

## "Color Scheme" Schema
* `primaryColor` - Displayed most frequently across your app.
* `primaryColorVariant`  - A tonal variation of primary color.
* `secondaryColor` - Accents select parts of your UI.
* `backgroundColor` - The underlying color of an app’s content.
* `errorColor` - The color used to indicate error status.
* `surfaceColor` - The color of surfaces such as cards, sheets, menus.
* `onBackgroundColor` - A color that passes accessibility guidelines for text/iconography when drawn on top of `backgroundColor`.
* `onPrimaryColor` - A color that passes accessibility guidelines for text/iconography when drawn on top of `primaryColor`.
* `onSecondaryColor` - A color that passes accessibility guidelines for text/iconography when drawn on top of `secondaryColor`.
* `onSurfaceColor` - A color that passes accessibility guidelines for text/iconography when drawn on top of `surfaceColor`.

**New Catalog Component**
![screen shot 2018-03-30 at 3 20 23 pm](https://user-images.githubusercontent.com/1753199/38150623-e3d0bc84-342d-11e8-8202-c2d5b87eeb11.png)

Commit:
[Themes] Add new semantic color scheme

The new color scheme includes error color and drops the lighter/darker
variants of secondary color.  This implementation should be drop-in compatible
for themers and other clients currently using MDCColorScheme, except for any
third-party code that might be using secondaryLightColor or
secondaryDarkColor.

randallli commented on components/Themes/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,81 @@
+#import <Foundation/Foundation.h>
+#import "MDCColorScheme.h"
+
+@protocol MDCSemanticColorScheming <MDCColorScheme>
```
Protocol docs.
We need to explain the difference between this and the MDCColorScheme protocol.

romoore replied:
Done

Commit:
Cleaning up API

romoore commented:
# Important!

Please don't merge until all reviewers have approved.

jverkoey commented on components/Themes/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,80 @@
+/**
+ Accents select parts of your UI.
+ */
+@property(nullable, readonly, nonatomic) UIColor *secondaryColor;
```
We have to decide where to implement the fact that `secondaryColor` falls back to `primaryColor` if null. A couple options I see available to us:

- [ ] We make this property `nonnull` on the protocol and `null_resettable` on the concrete class. We document that setting `secondaryColor` to nil will result in `secondaryColor` returning `primaryColor`. This means components blindly pull from `secondaryColor` without implementing the fallback logic.
- [ ] We keep the current approach and implement the fallback logic in each component. This means each component's code is more explicit about the behavior of choosing either secondary or primary.

What do you think?

jverkoey replied:
Another option:

- [ ] We make this property `nonnull` on the protocol and `nonnull` on the concrete class. We document in the concrete class that initializing with `secondaryColor` as nil will result in `secondaryColor` being assigned to `primaryColor`.

ianegordon replied:
I vote for option two.
initWithPrimary:(UIColor) // Set all the properties, including calculating a sane secondary
initWithPrimary:secondary:  // Set all the properties, based on the params

romoore replied:
Should the "secondary falls back to primary" be part of the protocol or not? Right now it's only in the concrete implementation.

@ianegordon I agree with @jverkoey that the new theme should not do any internal calculations that aren't explicitly documented. MDCBasicColorScheme *tries* to calculate new colors, but fails in many common cases.  For example, generating an MDCBasicColorScheme with only UIColor.red will result in the `primaryLightColor` being `UIColor.red`.  This breaks some components, notably [MDCBottomNavigationBar](https://github.com/material-components/material-components-ios/blob/develop/components/BottomNavigation/src/ColorThemer/MDCBottomNavigationBarColorThemer.m#L24) which will set both  `barTintColor` and `selectedItemTintColor` to `UIColor.red` as a result.

Personally, I'm not a fan of setting `secondaryColor` equal to `primaryColor`, but I understand it's a pretty firm requirement.

jverkoey replied:
> Should the "secondary falls back to primary" be part of the protocol or not? Right now it's only in the concrete implementation.

In the interest of less magic I think I also vote for #2, where the protocol has a nullable readonly secondary property and component themers must implement the fallback logic.

In other words: the current code is good, with the emphasis being that we need to ensure that all references to secondaryColor fall back to primaryColor in our themers.

romoore replied:
Why not make the color scheme implement the fallback?  That way it only needs to be done once and the themers can depend on `.secondaryColor` existing, even if it's identical to `.primaryColor`.

jverkoey replied:
Spoke offline with @romoore and we agreed on the following:

1. All of the protocol's properties should be readonly nonnull.
2. The concrete class implementation will not have any defaults or fallback logic - it will simply be a holder of the colors generated by the color tool. The color tool will provide defaults for all properties.

jverkoey commented on components/Themes/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,80 @@
+@property(nonnull, readonly, nonatomic) UIColor *primaryColorDarkVariant;
+
+/**
+ Accents select parts of your UI.
```
> If null, it is expected that the primaryColor will be used instead.

jverkoey replied:
This is assuming the API is still nullable, pending the decision below.

romoore replied:
Nullability wouldn't matter. If we set the properties to readonly, nonnull, then the implementations will need to ensure that all of the required set-up is done in `init`.

jverkoey commented on components/Themes/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,80 @@
+                     primaryColorDarkVariant:(nonnull UIColor *)primaryColorDarkVariant
+                              secondaryColor:(nullable UIColor *)secondaryColor
+                                  errorColor:(nullable UIColor *)errorColor;
+
```
Expose all of the properties on this class as writeable as well. It should be easy for someone to change the properties after initialization if they so desire (e.g. someone creates a Baseline color scheme initialized with the purples and then changes only the secondary color).

ianegordon replied:
Consider:
initWithPrimary:  // Calculate a sane secondary, error, etc.

I haven't seen this method of using nullable parameters as optionals before, typically I would expect all the parameters to be nonnull and there to be convenience constructors all wired into the designated init.

jverkoey replied:
We are intentionally not doing any code-side calculation of colors in the short term. That logic will live solely in the color tool.

romoore replied:
Regarding `readwrite` properties, do we have a specific use case where a client will need to modify a scheme once it has been created?  It seems to me that creating a new scheme as a variant of another is pretty easy.

```objectivec
  MDCSemanticColorScheme *colorScheme =
      [[MDCSemanticColorScheme alloc] initWithPrimaryColor:oldScheme.primaryColor
                                  primaryColorLightVariant:oldScheme.primaryColorLightVariant
                                   primaryColorDarkVariant:oldScheme.primaryColorDarkVariant
                                            secondaryColor:UIColor.magenta
                                                errorColor:[UIColor.red colorWithAlpha:0.5]];
```

jverkoey replied:
The primary use case I'm aware of is someone creating a Baseline color scheme and mutating its secondaryColor.

romoore replied:
I'm not sure what that would mean - do you imagine that the secondary color of the scheme would change midway during the lifecycle of an app of in the middle of user's workflow?

Are you thinking of scenario where it's not possible to create a new ColorScheme object and pass it along to all of the required consumers?  That the original has to be retained and then safely mutated?  If so, we should consider making the properties atomic.

jverkoey replied:
Spoke offline with @romoore and agreed that the property setter case may not be particularly important and that it may be sufficient to just have a designated initializer in the first pass of the API.

ianegordon commented on components/Themes/src/MDCSemanticColorScheme.m:
```diff
@@ -0,0 +1,63 @@
+    } else {
+      // Default is #ff1744
+      _errorColor = [[UIColor alloc] initWithRed:1
+                                           green:(float)(0x17 / 255.0)
```
Super-nit: CGFloat here and below.

romoore replied:
In light of #2931 (UIColor and NSCoding), I intentionally went with floats here so that components using this color value won't have to do gnarly work-arounds to check color equality.  An alternative (thanks @ajsecord), might be to place a category on UIColor:

```objectivec
+ (BOOL)mdc_isApproximatelyEqualToColor:(UIColor *)otherColor;
```

jverkoey replied:
The underlying data type for UIColor is CGFloat though, so anyone consuming a UIColor wouldn't be able to benefit from any float casting here (they can't know that we did and will have to assume the type is CGFloat).

romoore replied:
Can you clarify what you mean, @jverkoey ? As far as I understand from Apple's docs, color data is represented internally as a sequence of either 8-bit or 16-bit values (depending on RGB or P3 color spaces).  As far as I can determine, the CGFloat values are only used for equality testing.

jverkoey replied:
We'll likely remove the default from the code, so this discussion can be punted til the next time we implement hex colors in code. At that point, the decision will either be:

1. Document why a (float) cast is being used instead of a CGFloat.
2. Cast to CGFloat.

ianegordon commented on components/Themes/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,80 @@
+/**
+ A tonal variation of primary color.
+ */
+@property(nonnull, readonly, nonatomic) UIColor *primaryColorLightVariant;
```
Attaching light and dark to these variants will make it harder to implement darkMode.
If lightMode, statusBar.background = primaryColorLightVariant
else if darkMode, statusBar.background = primaryColorDarkVariant

statusBar.background = primaryColorEmphasisVariant

Naming is hard!

jverkoey replied:
My understanding presently is that these properties are semantically safe as light/dark variants regardless of the theme being light or dark, but I also have not seen any literal mappings be well defined in any componentry, so the jury's out on whether that's actually the case...

jverkoey replied:
At least for now, this is what our cross-platform spec is expecting as far as symbols.

romoore replied:
As far as I know, I don't think any of these are intended to be background colors.  Internally we are discussing about how far the "variants" should *vary* from the primary color, but it doesn't seem sufficient to support 4.5:1 contrast ratios.  I doubt the intention is for `.dark` to be background and `.light` to be text color (or *vice versa*)

jverkoey replied:
Correct, the dark and light variants are not intended to be used for overlaying text.

jverkoey replied:
We'll soon be adding onPrimary and onSecondary to the Color Scheme, which are accessibility-safe colors meant to be used for text and iconography overlaid on top of the primary or secondary color, respectively.

ianegordon replied:
My point is that LightVariant / DarkVariant aren't great from a semantic standpoint.
Simplifying to primaryColorVariant is more useful.  (Without light / dark)
statusBar.background = primaryColorVariant // Works in light or dark mode

jverkoey replied:
We'll be dropping the light/dark variants in favor of a single variant.

Commit:
Simplify the color scheme.

* <MDCSemanticColorScheming> no longer inherits from MDCColorScheme
* <MDCSemanticColorScheming> properties are now all nonnull.
* MDCSemanticColorScheme will no longer implement fall-back behavior. Clients
  should use the Material color tool to select appropriate colors.
* MDCSemanticColorScheme explicitly conforms to <MDCColorScheme> and
  <MDCSemanticColorScheming>

Commit:
Break-out MDCColorScheme compatibility

Making <MDCColorScheme> compatibility a category to keep the code a bit
cleaner. Adding basic -init unit test to ensure the properties are synthesized
and bound correctly.

Commit:
Add NSSecureCoding support.

Commit:
Adding extra test to ensure NSSecureCoding is supported.

Commit:
Documenting MDCSemanticColorScheme

ianegordon reviewed (approved):
I think we can safely mark my questions as "Out of Scope"
LGTM!
Please merge once CI is happy.

romoore commented:
Failing tests: 

```
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyDefault]' started.
components/TextFields/tests/unit/TextFieldControllerLegacyTests.swift:300: error: -[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyDefault] : XCTAssertEqual failed: ("82.0") is not equal to ("70.0") - 
components/TextFields/tests/unit/TextFieldControllerLegacyTests.swift:304: error: -[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyDefault] : XCTAssertTrue failed - 
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyDefault]' failed (0.022 seconds).
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyFullWidth]' started.
components/TextFields/tests/unit/TextFieldControllerLegacyTests.swift:322: error: -[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyFullWidth] : XCTAssertEqual failed: ("64.0") is not equal to ("57.0") - 
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyFullWidth]' failed (0.006 seconds).


Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldTests testSizing]' started.
components/TextFields/tests/unit/TextFieldTests.swift:235: error: -[components_TextFields_unit_test_swift_sources.TextFieldTests testSizing] : XCTAssertEqual failed: ("56.0") is not equal to ("49.0") - 
components/TextFields/tests/unit/TextFieldTests.swift:239: error: -[components_TextFields_unit_test_swift_sources.TextFieldTests testSizing] : XCTAssertEqual failed: ("80.0") is not equal to ("66.0") - 
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldTests testSizing]' failed (0.010 seconds).

//components/TextFields:unit_tests_IPHONE_7_PLUS_IN_10_3                 FAILED in 5.9s
```

romoore commented:
Rebased to latest develop. Hopefully that fixes the text fields failures.

jverkoey commented:
Note, we've now confirmed that the color scheme will include "on primary" and "on secondary". Feel free to add these in a follow-up PR if you prefer.

Commit:
Add onPrimary, onSecondary, background colors

romoore commented:
Text fields tests are the cause of the latest failures:

```
Test Case '-[components_TextFields_unit_test_swift_sources.MultilineTextFieldTests testSizing]' started.
components/TextFields/tests/unit/MultilineTextFieldTests.swift:158: error: -[components_TextFields_unit_test_swift_sources.MultilineTextFieldTests testSizing] : XCTAssertEqual failed: ("56.0") is not equal to ("49.0") - 
components/TextFields/tests/unit/MultilineTextFieldTests.swift:162: error: -[components_TextFields_unit_test_swift_sources.MultilineTextFieldTests testSizing] : XCTAssertEqual failed: ("80.0") is not equal to ("66.0") - 
Test Case '-[components_TextFields_unit_test_swift_sources.MultilineTextFieldTests testSizing]' failed (0.011 seconds).
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyDefault]' started.
components/TextFields/tests/unit/TextFieldControllerLegacyTests.swift:300: error: -[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyDefault] : XCTAssertEqual failed: ("82.0") is not equal to ("70.0") - 
components/TextFields/tests/unit/TextFieldControllerLegacyTests.swift:304: error: -[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyDefault] : XCTAssertTrue failed - 
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyDefault]' failed (0.022 seconds).
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyFullWidth]' started.
components/TextFields/tests/unit/TextFieldControllerLegacyTests.swift:322: error: -[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyFullWidth] : XCTAssertEqual failed: ("64.0") is not equal to ("57.0") - 
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldControllerDefaultLegacyTests testPresentationLegacyFullWidth]' failed (0.008 seconds).
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldTests testSizing]' started.
components/TextFields/tests/unit/TextFieldTests.swift:235: error: -[components_TextFields_unit_test_swift_sources.TextFieldTests testSizing] : XCTAssertEqual failed: ("56.0") is not equal to ("49.0") - 
components/TextFields/tests/unit/TextFieldTests.swift:239: error: -[components_TextFields_unit_test_swift_sources.TextFieldTests testSizing] : XCTAssertEqual failed: ("80.0") is not equal to ("66.0") - 
Test Case '-[components_TextFields_unit_test_swift_sources.TextFieldTests testSizing]' failed (0.003 seconds).

//components/TextFields:unit_tests_IPHONE_7_PLUS_IN_10_3                 FAILED in 5.1s
```

jverkoey commented:
Agreed with Ian's sentiment on the light/dark variant naming. Let's move the discussion to the color design doc and see if we can flatten the color scheme's semantics so that there's just a single primaryVariant.

jverkoey commented:
For now let's drop the light/dark variants from the API because they're not technically mapped to anything in iOS at this point.

jverkoey commented:
Remaining action items on this PR:

- [ ] Remove the primaryLightVariant and primaryDarkVariant APIs.
- [ ] Add a primaryVariant API.
- [ ] Add an onBackgroundColor API.
- [ ] Add a surfaceColor API.
- [ ] Add an onSurfaceColor API.

jverkoey commented:
Additional work:

- [ ] Expose a single init method for the Scheme type with no arguments.
- [ ] Initialize all of the parameters with the Material Defaults (documented [here](https://www.pivotaltracker.com/story/show/156171931)).
- [ ] Expose all of the Scheming properties as readwrites on the Scheme class.

We're still waiting for a team decision on whether we should expose a `+materialDefaults` API, so defer on adding that for now.

jverkoey commented on components/Themes/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,118 @@
+ A simple color scheme that provides more semantic context for the colors it uses. Unlike
+ @c MDCColorScheme, there are no optional properties, so more consistent theming can be applied.
+ */
+@protocol MDCSemanticColorScheming
```
`MDCColorScheming`

Commit:
Moving scheme to schemes/Color

Commit:
Merge branch 'develop' into new-colorscheme

Commit:
Moving Themes examples into schemes/Color

Commit:
Updating unit tests for Semantic color scheme

Commit:
Reverting ActivityIndicator changes.

Also rename the leave-behind header in Themes.

jverkoey commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,138 @@
+ A simple color scheme that provides semantic context for the colors it uses. There are no optional
+ properties and all colors must be provided, supporting more reliable color theming.
+ */
+@protocol MDCColorScheming
```
I suppose we don't technically need to implement NSObject here because all of the methods are required (no need for `respondsToSelector`), but can you confirm that the conformance was intentionally left out?

romoore replied:
I'm not sure, I tend not to force object requirements in protocols but I can add it. I have no feelings either way.

jverkoey replied:
Likely fine to leave as-is then until/if it becomes a problem :)

ianegordon replied:
Will clients need conformsToProtocol: if they are extending the protocol?

jverkoey replied:
If they are using conformsToProtocol: then I think they'll have already required a type that conforms to the NSObject protocol. Within our code and our themers I don't think we'll need to do any conformity or responder testing.

ianegordon replied:
SGTM.
We can add it if we need it.  Taking it away after we've added it, so much harder.

Commit:
Documentation FTW!

Commit:
Formatting FTW!

Commit:
Fixing supplemental files for schemes/Color

ianegordon commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,144 @@
+ A simple implementation of @c MDCColorScheming that provides Material baseline values from which
+ basic customizations can be made.
+ */
+@interface MDCSemanticColorScheme : NSObject <MDCColorScheming>
```
super-nit / naming is hard.
Can we assume all of our schemes will be semantic and call this MDCColorScheme?

romoore replied:
We cannot, since that's already a protocol name.

ianegordon commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,144 @@
+/**
+ Accents select parts of your UI.
+ */
+@property(nonnull, readonly, nonatomic) UIColor *secondaryColor;
```
secondaryColorVariant is listed in go/m2-color-overview.
Not here?

romoore replied:
It wasn't in pivotal, I leave the doc to the authors.

jverkoey replied:
We'll add these parameters in a follow-up pass once the parameter mappings are more clearly defined.

jverkoey commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -96,8 +96,14 @@
 - (nonnull instancetype)init;
 
+/**
+ Initializes an instance of MDCSemanticColorScheme with the Material baseline color values.
```
"Material defaults"

romoore replied:
Done

ianegordon commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,144 @@
+ @c backgroundColor.
+ */
+@property(nonnull, readonly, nonatomic) UIColor *onBackgroundColor;
+@end
```
onError?
go/m2-color-overview

romoore replied:
Again, not in pivotal, so I'll let the doc authors hash-out what's going to be built.

jverkoey replied:
We'll add these parameters in a follow-up pass once the parameter mappings are more clearly defined.

ianegordon commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,144 @@
+                            onSecondaryColor:(nonnull UIColor *)onSecondaryColor
+                              onSurfaceColor:(nonnull UIColor *)onSurfaceColor
+                           onBackgroundColor:(nonnull UIColor *)onBackgroundColor
+    NS_DESIGNATED_INITIALIZER;
```
Do we expect anyone to use this over init + properties?
What I would expect in the future is something like
- (instancetype)initWithJSONScheme: or the like.

If we keep it, please remember to add any add properties from above.

ianegordon commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,144 @@
+
+@end
+
+@interface MDCSemanticColorScheme (NSSecureCoding) <NSSecureCoding>
```
Why isn't this part of the standard declaration?
People (who are looking) will miss this and be frustrated.

romoore replied:
It's more Swifty, nobody complained during the last round of reviews so I kept it. If it's a big problem, I can move it.

ianegordon reviewed (changes requested):
Minor nits.
Very excited!

Commit:
Fixing BUILD files, imports, docs.

jverkoey commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -0,0 +1,144 @@
+                            onSecondaryColor:(nonnull UIColor *)onSecondaryColor
+                              onSurfaceColor:(nonnull UIColor *)onSurfaceColor
+                           onBackgroundColor:(nonnull UIColor *)onBackgroundColor
+    NS_DESIGNATED_INITIALIZER;
```
In the spirit of less APIs to maintain for now, let's remove this designated initializer for now and encourage use of the default initializer + property overrides.

romoore replied:
I thought this was being dropped?

jverkoey commented:
This PR has moved to https://github.com/material-components/material-components-ios/pull/3216 so that we can run CI against develop and land.
